### ISSUE TITLE
API is inconsistent on CPU vs. Processor

### ISSUE THREAD

**Issue by dbwiddis** [dblock/oshi#28]
Most modern machines have multi-core processors.  However, the API mixes terminology/code.  The Processor class has system-wide statistics, so should really only be instantiated once.  However, the HardwareAbstractionLayer refers to Processors in the plural, and implementations set up a loop to instantiate multiple (identical) copies of these virtual CPUs.

The multiple copies seem unnecessary with the current implementation.  However, there are some statistics, such as CPU load, which can be measured per-virtual CPU (although in the tick-counting between timestamps way, so iteratively calling all processors and waiting to count ticks is not a good model.)

Are we stuck with the current API or can we revise the structure?  And while we're at it, any chance we can avoid using the same class name as major JNA classes like "Memory"?

**Comment by dblock** [comment]
We can revise the structure/API by bumping the major version of the library and writing a proper UPGRADING doc.

**Comment by Underbalanced** [comment]
Additionally the CPU getProcessors()[val].getLoad() is highly inaccurate on multicore systems as this will only show that specific processor; 4 cores or more one core may register higher then the rest as such happens on my 32 core programming machine when testing this Library.

Further an issue, is that the current getLoad() already takes to much time; this would be alright if not having to loop through the processors and obtain an average CPU load correctly.

I am interested in contributing where possible as I find ways to improve it.

**Comment by dbwiddis** [comment]
Yes, the com.sun.management.OperatingSystemMXBean is pretty capable, and I only found oshi in my search to more accurately represent the available memory, which that MXBean (and oshi until my contribution) didn't model well.  If I were designing oshi from scratch I'd extend that MXBean...

Interestingly if you use that MXBean when you first start up the JVM, it doesn't accurately represent CPU usage. It needs a few microseconds to start collecting those tick averages.  We could consider a model where oshi retained tick state and updated the CPU % with each successive call.  Or we could calculate the load average and divide by number of cores and get pretty close, too.   Unfortunately, these are implementation choices and it should not fall on us to specify which version the user should use; we should provide a means to all available information.

Of course, Windows doesn't have all the same info that Linux and Mac do.  The bottom line is we should decide how we want to represent these things, and do so consistently and usefully (particularly in the way of providing functionality that does not exist elsewhere.)

**Comment by Underbalanced** [comment]
Additionally, I think that we should include an Event interface of some sort that can be written into the method, so that one can either include an Event Handler or by default an empty one would be created, this would allow one to create a thread, scheduled or otherwise and interface into the output, allowing it displayed or fed else where. Just an idea.

**Comment by dbwiddis** [comment]
I defer to your expertise on the Event interface. Sounds like a good idea!

Current pull request adding processor speed (MHz/GHz) also has different options (max vs. current, which can vary per-core) that we should probably include in this conversation.

**Comment by Underbalanced** [comment]
A basic Gist.

```
public class FunctionalUtils {

    /**
     * This will set an EventNotifier, if the EventNotifier supplied is null, an empty
     * EventNotifier will be created to prevent NullPointerExceptions.
     *
     * @param en EventNotifier representing the EventNotifier supplied
     * @return EventNotifier representing the Non null EventNotifier supplied or a Blank one if null was provided
     */
    public static EventNotifier setNotifier(EventNotifier en) {
        if (en != null) {
            return en;
        } else {
            return FunctionalUtils::handle;
        }
    }

    private static void handle(Object k, Object d){}
}
```

&

```
/**
 * Represents a interface used in supplying to generic libraries with a nexus for event handling.
 * The event handler can be implement in any fashion. When creating a Object requires this Event
 * Notify Handler means that that object will be publishing a Object Key and Object data. When
 * implementing this interface you will be fed the data from the event and be able to implement
 * any style event handler.
 *
 * @author Daniel (Underbalanced) (Free Universe LLC. - www.free-universe.com)
 * @since 1.0.0.0
 */
@FunctionalInterface
public interface EventNotifier<KEY, VALUE> {
    void eventNotify(KEY key, VALUE data);
}
```

Here we will create a SystemInformation class
```
SystemInformation si = new SystemInformation(eventNotifier); // optional param

or

SystemInformation si = new SystemInformation(new EventNotifer<Object, Object>(){
    void eventNotify(Object key, Object data){
                 /* this allows the implimentor proccess the data which is useful if we say made a thread that gathers data at a interval */
     }
});

or you can use a method reference or new class.
```

Now insdie of SystemInformation
```
We would set an Event Notifier field using FunctionalUtils.setNotifier and in the method for getting data...

public double cpuLoad(){
          ....
          double cpuLoad = blah;
          EVENT_NOTIFIER.eventNotify("CPU_LOAD", cpuLoad);
}

```
I would have to push this to a public library for better examples, currently its in my private repos, where i use this style to maybe implement further logic on something that already has some logic in a modular, non code breaking way.

**Comment by dbwiddis** [comment]
I've just spent about an hour reading a ton of information on this subject and propose the following terminology:
Processor = One single processor per machine. Contains common resources, e.g., memory, overall load factor, vendor id/name/rated clock frequency.
CPU = Multiple physical and/or logical processors per machine. Can be cores, or multiple processors per core. (My quad-core MacBook Pro has 4 cores but 8 virtual CPUs).   Each CPU has its own current frequency and CPU usage percentage.

**Comment by dbwiddis** [comment]
Here's where I'd like to go with this for the next release (probably 2.0 since my vision will break the current API):
* Existing "Processor" class will have only one per Hardware Abstraction Layer.
  * Replace current CPU Load implementations
    * Current implementations flawed
      * Mac OS X requires a delay to count between ticks; return the raw tick values instead
      * Linux requires a delay to count between ticks that is not yet implemented (it measures from startup); return the raw tick values instead
      * Windows requires parsing `wmic` shell command; investigate ways to get ticks programmatically
    * Replace with `com.sun.management.OperatingSystemMXBean` `getSystemCpuLoad()`
      * This requires Java 1.7.  I think it's worth it to overcome the obstacles above.
  * Add System Load average for Mac OS X and Linux
    * `OperatingSystemMXBean` `getSystemLoadAverage()`
      * This requires Java 1.6
      * Returns -1 for Windows as it's not applicable
* Processor class will have an Array(List?) of Logical CPU classes
  * Return actual processor frequency per-logical CPU
  * Return actual ticks per-logical CPU (calculate %Load if possible)
* Allow user to create a separate thread (probably a `ScheduledExecutorService`) to periodically (per user specification) get the tick values, calculate %CPU, create events, etc.

**Comment by dbwiddis** [comment]
I think I've got a way to update the Processor without breaking the current API.  The Processor objects can still represent Logical CPUs as they currently do, but will just have getters and setters for the common info, which will be calculated once outside the objects when they are created.

Still want to go to 1.7 for the `getSystemCpuLoad()` method. We could do some tick-counting magic with a background thread to work around it, but that's a lot of work to bypass a single line of code.

**Comment by dbwiddis** [comment]
Well, it doesn't look like the `OperatingSystemMXBean` is going to be helpful for CPU Load. Numbers only appear to be nonzero for Linux -- and are still probably wrong.  (And I get all sorts of compiler warnings about using the com.sun version of the MXBean -- probably because it's so wrong!) I think I read somewhere that the MXBean needs to be running for some amount of time to be useful.

Load average works (and is official Java API as of 1.6), but Windows doesn't have a load average.  

Here's a comparison of the current (counting ticks) method vs. the MXBean methods for CPU (which should match) and Load Average (which should, when divided by number of CPUs, approximate CPU %):

For mac, the CPU% shows 0:
```
CPU load (current method): 6.6831684%
CPU load (OperatingSystemMXBean): 0.0%
CPU load average: 1.41552734375
CPU load estimated from load average: 17.694091796875%
```
For Linux, the current method is wrong (it only counts from 0, rather than between-ticks).  The MXBean appears to match, so is equally useless... I was intentionally overloading my CPU (load factor of 5 on a 1-core machine) and the numbers never budged.
```
CPU load (current method): 0.35%
CPU load (OperatingSystemMXBean): 0.3654685938521952%
CPU load average: 5.04
CPU load estimated from load average: 100.0%
```
On Windows, CPU from the MXBean is 0%, and of course, Windows doesn't have load average so the MXBean returns -1 (as expected):
```
CPU load (current method): 32.0%
CPU load (OperatingSystemMXBean): 0.0%
CPU load average: -1.0
CPU load estimated from load average: -100.0%
```

**Comment by dbwiddis** [comment]
AHA.... mystery solved.  The first time CPU load is called it always returns 0 (this apparently initializes it).  It has to be called a second time to get the information....

**Comment by dbwiddis** [comment]
And it needs a half-second or so after initialization to begin to return reasonable results -- same as our tick-measuring method, but can be doing other things in the program during that time.

**Comment by dbwiddis** [comment]
#51 submitted to address
* Implements `OperatingSystemMXBean` for CPU Load (requires initializing and waiting a bit to generate reliable results but can be repeatedly polled later with no delay) and CPU Load Average (Mac/Linux).
* Deprecated the old getLoad() and fixed the Linux implementation to do the same tick-counting/delay that the Mac implementation did.  Windows still command-lines wmic.

Still TODO:  Expose `getTicks()` once I figure out how to get that info on Windows. Also I know how to get per-core ticks on Mac/Linux and can create a new `getTicksPerCpu()` method, but looking for how to get that info on Windows as well.  I believe it may be buried in the [Performance Counters](https://msdn.microsoft.com/en-us/library/windows/desktop/aa373083(v=vs.85).aspx) somewhere.

**Comment by dbwiddis** [comment]
Found a way to get overall ticks on Windows: http://www.codeproject.com/Articles/9113/Get-CPU-Usage-with-GetSystemTimes

**Comment by dbwiddis** [comment]
Updated #51 with tick counts!  All that's left is per-core ticks.

**Comment by dbwiddis** [comment]
My plan to address the processor issue:  I'll make a new object, SystemProcessor, representing systemwide statistics (all of the existing ones). The existing Processor objects (CentralProcessor) will contain a single static object for the System version, and simply pass-through the existing methods to the static object, so we'll only ever have one.  That will free us up to implement per-core methods.

The SingleProcessor object will need an index number to identify themselves and pick the correct row out of per-core stats.  I'll make a constructor with that index number for future use and deprecate the existing no-arg constructor -- which will call a synchronized incrementing counter in the SystemProcessor object.

**Comment by dbwiddis** [comment]
Resolved in #62 without breaking API
